### ISSUE TITLE
JSON.parse() performance

### ISSUE THREAD

**Issue by scarlac** [facebook/hermes#811]
## Bug Description

TL;DR: `JSON.parse()` is 3x slower on Hermes than JSC.

```
| Engine                  | Method                 | Result |
|-------------------------+------------------------+--------|
| Hermes v0.12.0          | parseJson median       |   50.0 |
| Hermes v0.12.0          | parseJson avg          |   54.5 |
| JSC                     | parseJson median       |   15.0 |
| JSC                     | parseJson avg          |   16.6 |
| NodeJS v16.13.2         | parseJson median       |   15.0 |
| NodeJS v16.13.2         | parseJson avg          |   15.4 |
```

Porting a React Native app to use Hermes can cause performance regressions depending on use case.
I have narrowed it down to large JSON payloads being loaded. Specifically, `JSON.parse()` is much slower in Hermes than JSC. JSC is 70% faster than Hermes / Hermes is ~3x slower than JSC. This is significant for our use case and blocks us from migrating.

It's probably fair to say that parsing JSON data from a string is a very common use case. The performance regression (compared to JSC) is severe enough that we cannot use Hermes in production. We have tried 3 times and gotten complaints from users.

- [x] I have run `gradle clean` and confirmed this bug does not occur with JSC

Hermes version: v0.12.0
React Native version (if any): 0.68.2
OS version (if any): Android 12.0
Platform (most likely one of arm64-v8a, armeabi-v7a, x86, x86_64): arm64-v8a

## Steps To Reproduce

1. Call `JSON.parse(largePayloadString)`
2. Compare with JavaScriptCore and NodeJS

## Code Example:
https://github.com/scarlac/benchmark-fs#command-line-test-cli-testjs

The repo contains several performance tests but for the purpose of this ticket, we should focus on the cli-test.js file which can be run with the Hermes CLI tool:
`hermes cli-test.js`

NodeJS performs similar to JSC. It'd be amazing to have Hermes at that level as well.

## The Expected Behavior
Performance should be similar (🤞 or better) than JSC.

**Comment by tmikov** [comment]
I was able to repro this with the file at the link.

**Comment by fbmal7** [comment]
Hi @scarlac, just wanted to update you that the team has started an investigation into this issue. We are optimistic that we can bring perf improvements to JSON.parse, to bring it on par to JSC/V8. We will be entering the experimentation phase soon to see what changes help. We will update this thread as we make progress.

I wanted to ask more about the kinds of JSON data you are actually processing. In the benchmark you provided, objects with the same keys & values are being constructed many times. Is that representative of your actual production workload? Do you expect that a lot of the data you are processing will have the same keys or values? If there's any other useful benchmarking tests for JSON parsing you can share, that would be helpful as well.

**Comment by scarlac** [comment]
@fbmal7 love to hear it. The data we typically parse is more varied and deeper nested but I wanted to make the test case clear and simple to begin with.

I’m reality we see a lot of variety in the values. Eg GUIDs, long lists of photos with local file paths (formatted as a URI), metadata with numbers and strings, product names, etc.

**Comment by billouboq** [comment]
@fbmal7 One big use case for this is the usage of redux-persist which save and load data from JS to any database we want. This is done a lot of times even if it's throttled

**Comment by fbmal7** [comment]
@scarlac We just landed [40de6b5](https://github.com/facebook/hermes/commit/40de6b54636eee41c3a0f347f9144f5b94e42a5e) which did bring a decent perf win. When I ran it locally against the benchmark you provided, it improved the speed by 33%. We still have a long way to go to match the perf of V8/JSC, but this was an easier optimization to land first.

**Comment by fbmal7** [comment]
We don't tend to cut releases for Hermes very often. Building Hermes from source is always an option, and it's not too complicated. If you just want to build hermes and run it on the command line, you can use [these](https://github.com/facebook/hermes/blob/eabf5fcd25b3eefbfa8af694c72ad78d7a9a48c7/doc/BuildingAndRunning.md) instructions.

**Comment by scarlac** [comment]
For those wondering, it looks like RN 0.71.0 is currently targeting Hermes tag `hermes-2022-11-03-RNv0.71.0-85613e1f9d1216f2cce7e54604be46057092939d` which include the 33% improvement.

**Comment by fbmal7** [comment]
Hi @scarlac, just wanted to update you again. To get better improvements past the simple one we already landed has been a lot more work than initially anticipated. However, I wanted to let you know the team is still looking at this issue! We are exploring optimizations such as:

- Making the parser iterative
- Parsing strings for object properties vs. all other strings differently
- Minimizing the amount of times we loop over scanned strings
- Creating a new internal fast path API to set the property on the JSObject we are building during parse

I'm not too sure on the timeline for all of this, but will keep the thread updated as we make progress.

**Comment by scarlac** [comment]
Thanks @fbmal7 . I appreciate the details and the teams hard work on this! Sounds like it'll make for some great release notes on the RN blog :)

**Comment by fbmal7** [comment]
@scarlac My favorite thread to update! I'm finally getting around to landing some of the changes I wrote a couple months back. Two optimizations just landed yesterday and will make their way into 0.72. They did not bring as big of a win as the previous optimization.

- [Incrementally calculate string hash](https://github.com/facebook/hermes/commit/de9cff2aa41fc1f297b568848143347823d73659), ~8% win
- [Precompute if a string is ASCII or not](https://github.com/facebook/hermes/commit/6e2dd652c8d90c5d59737a81f66a259efffdcd00), ~4% win

The next change I have planned is one of the ones I referenced above- only performing expensive hash computation+lookups on strings that are object keys. Currently we are always performing a hash lookup for all strings we are scanning in the parser. We don't differentiate parsing for string keys vs. string values.

**Comment by scarlac** [comment]
@fbmal7 Love getting these updates. Really impactful for everyone using RN. Thanks again!

**Comment by henrymoulton** [comment]
https://twitter.com/radexp/status/1630581333318680577

@radex would be great to hear about your JSON parsing improvements :)

**Comment by radex** [comment]
yes, yes, soon :)

**Comment by radex** [comment]
https://github.com/facebook/hermes/pull/933

**Comment by henrymoulton** [comment]
Hey @fbmal7, the blog post from @radex was a really amazing deep dive into the JSON parsing improvements. Although totally optional, it'd be interesting to hear any plans you have off the back of #933 to improve it.

**Comment by darwinshameran** [comment]
We're seeing the same issue, reverted back to using JSC for now. Would love to see Hermes performance being on par with JSC.

**Comment by hakansaglam29** [comment]
We upgraded react native version from 0.69.5 to 0.71.13 and then we just started to get syntax error from JSON.parse(). After deactivating hermes again, the issue is resolved. We could not use hermes in production!

**Comment by scarlac** [comment]
> We upgraded react native version from 0.69.5 to 0.71.13 and then we just started to get syntax error from JSON.parse(). After deactivating hermes again, the issue is resolved. We could not use hermes in production!

This doesn't sound related to the performance issue. Can you specify what exactly broke and provide a json snippet or at least the error message? Otherwise it's not possible to investigate.

**Comment by hakansaglam29** [comment]
> > We upgraded react native version from 0.69.5 to 0.71.13 and then we just started to get syntax error from JSON.parse(). After deactivating hermes again, the issue is resolved. We could not use hermes in production!
> 
> This doesn't sound related to the performance issue. Can you specify what exactly broke and provide a json snippet or at least the error message? Otherwise it's not possible to investigate.

This is not a performance issue but I think it should be resolved together.

We were trying to parse a data like this: 

"{\"lorem\":{\"xx_version\":\"1.0.0.0\",\"yy_version\":\"1.0.0.1\",\"zz_build\":\"Jul 17 2022 14:35:23\",\"kk_version\":\"1\",\"hh_id\":\"tttt-b-c8a952066002\",\"gg_id\":\"@xxxxxxx\",\"cc_id\":\"686725B847A0\"}}"

After switching to new react native version 0.71.13, we just started to get SyntaxError: Unexpected token ..., although it is in the appropriate format.

**Comment by tmikov** [comment]
@hakansaglam29 I just tried your example and it parses with Hermes without an error.

**Comment by hakansaglam29** [comment]
> @hakansaglam29 I just tried your example and it parses with Hermes without an error.

Yes. if you try this output directly, JSON.parse is working. But i am getting this output as a response and using  with  JSON. parse. After that it throws an error. I could not understand why. But somehow after disabling Hermes, it worked again.

**Comment by fbmal7** [comment]
@hakansaglam29 We can't debug the issue until you provide working repro steps. You should also open a new issue as this doesn't have to do with JSON parsing performance. There could be other polyfills or changes that happen when you switch to Hermes that are actually the root cause of your issue, not `JSON.parse`.

**Comment by hakansaglam29** [comment]
> @hakansaglam29 We can't debug the issue until you provide working repro steps. You should also open a new issue as this doesn't have to do with JSON parsing performance. There could be other polyfills or changes that happen when you switch to Hermes that are actually the root cause of your issue, not `JSON.parse`.

Thanks @fbmal7 i will open a new issue.

**Comment by scarlac** [comment]
Thank you so much for the improvements to Hermes since we opened this.
While Radex provided a SIMD example I understand we can't just use that.
However, the still-very-slow JSON processing discrepancy is preventing Tesla from fully migrating to Hermes. Given that Reanimated seems to have abandoned JSC support for development, and everything depends on it, we are very motivated to migrate, but the performance hit is still too significant.

Here is the latest benchmark on the same M1 as I used originally.

## Hermes (built from tag `hermes-2025-01-13-RNv0.78.0-a942ef374897d85da38e9c8904574f8376555388`)
```
9:42:30 ~/jobs/benchmark-fs ⚡️ ./hermes cli-test.js 
[...]
parseJson median 43
parseJson avg 42.7
```

# NodeJS v18.12.1 (Node 20 is actually a bit faster than this!)
```
9:42:37 ~/jobs/benchmark-fs ⚡️ node cli-test.js 
[...]
parseJson median 14
parseJson avg 15.4
```

Is there anything we can do to encourage further improvements?

**Comment by tmikov** [comment]
@scarlac what would be acceptable performance? We can't necessarily match v8, because we have much more severe size constraints: simdutf is about 200K, which is very high for us (though perhaps we don't need all of it).

**Comment by scarlac** [comment]
Is it a JIT limitation? I don't know much about SIMD myself so I'll take your word for it, although I think it could arguably be worth it if all payloads see significant improvements.

If we can't get parity then I would say my threshold for not worrying about it would be around 25ms vs NodeJS at 15ms.
At that point we don't have to rearchitect everything but just do standard optimizations here and there.

**Comment by SimpleCreations** [comment]
My 2 cents here — perhaps it's possible to include the SIMD implementation behind an optional build flag. I'm sure many app developers would gladly accept 200K app size penalty in favor of 3-4x JSON performance improvement

**Comment by scarlac** [comment]
@tmikov Do you think it's something we could hope to get included in Hermes behind an experimental flag? It looks like there is reasonable interest in such a thing and I don't think I'm understating the importance of fast JSON parsing here - it's literally one of the most compute intensive operations many mobile apps will be doing, both from server requests, persisted storage data parsing, etc.

**Comment by scarlac** [comment]
@tmikov I've re-run the benchmark with the latest version of Hermes on a new M4 Pro, given that JSC is being forced out of RN and the community at the moment:

https://github.com/scarlac/benchmark-fs/blob/main/README.md#results-m4-pro

| Engine                  | Method                 | Result |
| ----------------------- | ---------------------- | ------ |
| Hermes 2025-09-01       | parseJson median       |   26.0 |
| Hermes 2025-09-01       | parseJson avg          |   28.6 |
| NodeJS v20.19.4         | parseJson median       |    9.0 |
| NodeJS v20.19.4         | parseJson avg          |   10.0 |
| JSC                     | parseJson median       |    6.0 |
| JSC                     | parseJson avg          |    7.3 |
| Deno                    | parseJson median       |    8.0 |
| Deno                    | parseJson avg          |    8.8 |

Conclusion is that Hermes is still ~3x slower than everyone else.
Is there anything we can do from Tesla's side to encourage further improvements in this area? I am unfortuantely not personally strong with C++.

**Comment by tmikov** [comment]
@scarlac I completely agree - this is really bad and must be addressed. The challenge is that the Hermes team has to focus on Meta's internal priorities, which at this time do not include JSON parsing. We don't have complete freedom to choose what to work on. I have been trying, alas unsuccessfully, to find a way internally to raise the priority of JSON parsing performance.

To some extent these priorities are determined by the RN teams and maintainers, so I think our best hope for addressing this is convincing them that this is a critical issue.

I know this is not ideal.

Meanwhile, when I have free time, I will do some benchmarking - 3x difference is so dramatic that it should be easy to get at least some wins.

**Comment by fbmal7** [comment]
@scarlac I'm excited to share that we have some pretty big updates underway to our JSON parser. It should be merged to our `static_h` branch within a week or so. On the repo you shared, these changes bring a speedup of ~2.5x.

How representative of your actual JSON workload is the sample you shared? Any other potential sources of tests would be helpful.

**Comment by scarlac** [comment]
Great news! Very excited to give it a spin.

> How representative of your actual JSON workload is the sample you shared? Any other potential sources of tests would be helpful.

We have a wide variety. But generally speaking it's very large arrays of semi large objects. There is a small amount of nesting but I will try and find some samples and update the benchmark so we can compare. 

Thank you!

**Comment by fbmal7** [comment]
@scarlac the team has a big update to share on `JSON.parse`! Over the past several weeks we've done a complete overhaul of the way the JSON parser works. This landed over 11 commits from [6d166c239e850160195ffb6fddbca11cdd5f2d1f .. 22d5252cf0a4a29a8929fc58154a508f89228443]. The title of each of the commits provides a good summary, but here is an even more condensed version:

- Remove abstraction layers. The JSONLexer operates directly on raw pointers, instead of a streaming API that abstracted over the input string's encoding (UTF8 / UTF16 / ASCII.)
- Switch from a recursive to iterative parser. This allows for creating VM structures with known sizes upfront.
- Cache object hidden classes for parsed objects, keyed by depth.
- Refactor the code for lexing a string to optimize for plain strings and strings containing escapes separately.
- Replace our existing number parser with a much more optimized library for parsing numbers (fast_float.)
- Add an even faster number path for integers that does not call into fast_float.
- Avoid all allocations for lexing numbers.
- Use a lookup table to quickly determine the type of token for a given ASCII character (e.g., one index into this table tells you if a character is a whitespace character.)
- Use a dispatch table in the core loop of the lexer based off the values of the above lookup table.

This new code was benchmarked across a wide variety of JSON workloads. The original snippet that was shared on this issue now runs ~2.7x faster. On some number heavy workloads that consist primarily of floats, we achieved a speedup of ~3.41x.

**Comment by scarlac** [comment]
This is great news, @fbmal7 ! I'll give it a spin. Thank you so much to you, @tmikov and the team!

**Comment by radex** [comment]
Amazing stuff @fbmal7!

**Comment by billouboq** [comment]
That's amazing ! 💪 
So much people will benefit from that

**Comment by mobinni** [comment]
This is amazing, what version of react-native are these targeted for?

**Comment by scarlac** [comment]
@fbmal7 @tmikov 
Massive improvement! With the latest commit on `static_h` (d9fba48):
**65-66% reduction regardless of which generation Apple silicon is used.**

The actual results:

# M1 Pro
```
../hermes_workingdir/build_release/bin/hermes cli-test.js 
parseJson, ms:  37
parseJson, ms:  28
parseJson, ms:  18
parseJson, ms:  17
parseJson, ms:  16
parseJson, ms:  17
parseJson, ms:  16
parseJson, ms:  16
parseJson, ms:  17
parseJson, ms:  17
parseJson median 17
parseJson avg 19.9
```

# M4 Pro
```
../hermes_workingdir/build_release/bin/hermes cli-test.js 
parseJson, ms:  21
parseJson, ms:  13
parseJson, ms:  17
parseJson, ms:  11
parseJson, ms:  13
parseJson, ms:  9
parseJson, ms:  11
parseJson, ms:  10
parseJson, ms:  9
parseJson, ms:  10
parseJson median 9
parseJson avg 12.4
done
```

This brings Hermes much closer to the competition.

**Comment by scarlac** [comment]
@tmikov @fbmal7 
You asked for a more specific example. Here's a payload generator based on actual data (strings, etc. are anonymized/randomized but lengths are kept about the same).
https://github.com/scarlac/benchmark-fs/blob/main/cli-test2.js

You can see the the new test is slightly tougher, but the improvements hold really well. On an M4 Pro, I see the following for the `static_h` branch:
```
11:57:08 ~/benchmark-fs ⚡️ ../hermes_workingdir/build_release/bin/hermes cli-test2.js                                  
generating dummyData via method 2 (nested task lists) to reach ~8 MB...
dummyData template size, in kB: 6
dummyData size in kB: 8192
parseJson, ms:  17
parseJson, ms:  17
parseJson, ms:  17
parseJson, ms:  17
parseJson, ms:  17
parseJson, ms:  16
parseJson, ms:  17
parseJson, ms:  17
parseJson, ms:  17
parseJson, ms:  16
parseJson median 16
parseJson avg 16.8
done
```

And the following for the `hermes-2025-07-07-RNv0.81.0-e0fc67142ec0763c6b6153ca2bf96df815539782` tagged build:
```
11:59:12 ~/benchmark-fs ⚡️ ../hermes_workingdir_stable/build_release/bin/hermes cli-test2.js
generating dummyData via method 2 (nested task lists) to reach ~8 MB...
dummyData template size, in kB: 6
dummyData size in kB: 8192
parseJson, ms:  45
parseJson, ms:  38
parseJson, ms:  35
parseJson, ms:  37
parseJson, ms:  36
parseJson, ms:  36
parseJson, ms:  36
parseJson, ms:  36
parseJson, ms:  36
parseJson, ms:  38
parseJson median 36
parseJson avg 37.3
done
```

**Comment by fbmal7** [comment]
@scarlac Thanks for the additional data.
> strings, etc. are anonymized/randomized but lengths are kept about the same

You should be careful about how you do this. On your personal machine it's worth testing against the actual real payload. If you randomize all the keys of objects you could be severely hampering an optimization that was landed in fa3e54bc384bab1400e0137cb081f3ee23f975a8. We try to reuse the shape of objects that were found earlier on in parsing, which can be quite a common pattern in JSON payloads.

edit- I checked the repo and see that you are ensuring stable object keys, so the above warning should not apply.

**Comment by scarlac** [comment]
@fbmal7 Yeah, that's why I added `randomEnum()` from a limited list of items, so there's a reasonable statistical chance that strings are reused according to the frequency you might see in our original data. It's not 1:1, but what's most important here is the delta. It's still great.
I also retained the original benchmark script 100% as-is, so anyone can reliably test the progress over time.

Regardless of the script/payload I use, I see around 1.5x slower performance on Hermes, but that's still a huge improvement compared to the current stable version (3-4x slower). So the optimizations you guys made will be a leap forward for all RN apps across the globe.

**Comment by ksemenov-faraway** [comment]
Hi guys! Starting from which version of RN this improvements included?

**Comment by SimpleCreations** [comment]
I really wish there was more transparency from the Hermes team about the releases 🥲

The Releases section is not updated, there are no changelogs and it's not clear what Hermes changes are included in each RN release

**Comment by scarlac** [comment]
@ksemenov-faraway 

The easiest answer to this is RN 0.84 - It was announced here:
https://reactnative.dev/blog/2026/02/11/react-native-0.84#hermes-v1-as-default

The slightly more involved answers are, since RN 0.83:
https://reactnative.dev/blog/2025/12/10/react-native-0.83#experimental---hermes-v1

And, you can include Hermes v1 in an older version of RN:
https://github.com/facebook/hermes/blob/static_h/doc/ReactNativeIntegration.md#using-a-custom-hermes-build-in-a-react-native-app

And, the version of Hermes in any React Native release/tag (in this case v0.84):
https://github.com/facebook/react-native/blob/v0.84.0/packages/react-native/sdks/hermes-engine/version.properties

Since Hermes is maintained separately from RN, it's the RN project that decides what version it wants to use. So even if Hermes did releases, it would not be apparent what version is used.
